scipy/linalg/decomp.py

```diff
@@ -44,28 +44,53 @@ def _make_complex_eigvecs(w, vin, dtype):
     return v
 
 
-def _geneig(a1, b1, left, right, overwrite_a, overwrite_b):
+def _make_eigvals(alpha, beta, homogeneous_eigvals):
+    if homogeneous_eigvals:
+        if beta is None:
+            return numpy.vstack((alpha, numpy.ones_like(alpha)))
+        else:
+            return numpy.vstack((alpha, beta))
+    else:
+        if beta is None:
+            return alpha
+        else:
+            w = numpy.empty_like(alpha)
+            alpha_zero = (alpha == 0)
+            beta_zero = (beta == 0)
+            beta_nonzero = ~beta_zero
+            w[beta_nonzero] = alpha[beta_nonzero]/beta[beta_nonzero]
+            if numpy.iscomplexobj(alpha):
+                w[~alpha_zero & beta_zero] = complex(numpy.inf, 0)
+                w[alpha_zero & beta_zero] = complex(numpy.nan, numpy.nan)
+            else:
+                w[~alpha_zero & beta_zero] = numpy.inf
```

Comment:
Branch not covered by tests, looks ok though (cf. `python runtests.py -s linalg --coverage; firefox build/coverage/index.html` or the codecov link on this page) -- only commenting since below there's a claim that all branches are covered ;)